### ISSUE TITLE
Strict equality not working between string and uuid

### ISSUE THREAD

**Issue by sk-** [python/mypy#18798]
**Bug Report**

`--strict-equality` seems to only work in very reduced cases. In particular, it is not working when comparing strings and uuids.

**To Reproduce**

```python
import uuid
from typing import TYPE_CHECKING, reveal_type

a = uuid.uuid4()
b = str(a)

if TYPE_CHECKING:
  reveal_type(a)
  reveal_type(b)
print(a == b)

c = "c"
d = 500
print(c == d)
```
https://mypy-play.net/?mypy=latest&python=3.12&flags=show-error-codes%2Cstrict-equality&gist=6c8273316b52930967ad0f8062c57aa3

**Expected Behavior**

The comparison in line 10 should raise a comparison overlap error. Something like

```
test_strict_equality.py:10: error: Non-overlapping equality check (left operand type: "str", right operand type: "uuid.UUID")  [comparison-overlap]
```

**Actual Behavior**

Mypy does not flag the string/UUID comparison as incompatible

```
test_strict_equality.py:8: note: Revealed type is "uuid.UUID"
test_strict_equality.py:9: note: Revealed type is "builtins.str"
test_strict_equality.py:14: error: Non-overlapping equality check (left operand type: "str", right operand type: "int")  [comparison-overlap]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.15.0
- Mypy command-line flags: `--strict-equality --show-error-codes`
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.12.9

**Comment by sterliakov** [comment]
It only works when there's no custom `__eq__` on types of both sides:

https://github.com/python/mypy/blob/e37d92d6c2d1de92e74c365ee1240c67c94c24b3/mypy/checkexpr.py#L3760-L3761

Any type can compare equal to anything (`def __eq__(self, other): return True`), and it isn't recognizeable just by looking at `__eq__` signature. `uuid.UUID` has custom `__eq__` defined in [typeshed](https://github.com/python/mypy/blob/e37d92d6c2d1de92e74c365ee1240c67c94c24b3/mypy/typeshed/stdlib/uuid.pyi#L61). 

However, AFAIC that method only compares equal to other UUIDs, so perhaps it should be excluded from typeshed?

**Comment by sterliakov** [comment]
No, current typeshed consistently adds `__eq__` to all classes that redefine it, so changing this for `UUID` would be unprecedented.

**Comment by sk-** [comment]
@sterliakov This is really unfortunate, because the code is [special casing](https://github.com/python/mypy/blob/bbd7a6cf5886ac7162f0f1cce5c34e89c807262e/mypy/typeops.py#L1083) base types. That's why it works for strings and lists, as both the typings for [strings](https://github.com/python/typeshed/blob/0b8dcfc00b06b045f82cb10a22ae52b613a939db/stdlib/builtins.pyi#L597) and [lists](https://github.com/python/typeshed/blob/0b8dcfc00b06b045f82cb10a22ae52b613a939db/stdlib/builtins.pyi#L1094) also have a custom declared `__eq__` method. It can be be seen in this [playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=show-error-codes%2Cstrict-equality&gist=5801c3bba58a0f5daa340da4467e7dc7) that mypy correctly reports a non overlapping error when checking them for equality.

I would argue against the decision of closing this issue, because `uuid.UUID` is a type from the standard library and could be easily special cased, just like builtin types.

If this is in fact is the intended behavior, could you please add a note to the documentation https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-strict-equality.

**Comment by sterliakov** [comment]
I'm slightly against introducing more special cases for stdlib like this - mypy code is fairly non-trivial, and this issue can be solved at the stubs/typeshed level more efficiently in checker-agnostic fashion. Pyright suffers from exactly the same problem now.

Perhaps you could raise this issue (slightly rephrasing to make it clear that custom `__eq__` is harmful for `uuid.UUID` typing) in [python/typeshed](https://github.com/python/typeshed) instead? Since we [know](https://github.com/python/cpython/blob/55815a6474c59001f0230e44560341b643268e87/Lib/uuid.py#L283) that UUIDs only compare equal to other UUIDs, we may lie in stubs and omit custom `UUID.__eq__`  entirely. @hauntsaninja I'm not perfectly familiar with typeshed guidelines - is such an omission feasible?
